tests/legacy-cli/e2e/tests/build/server-rendering/server-routes-preload-links.ts

```diff
@@ -0,0 +1,223 @@
+          }
+      `,
+    ),
+  ]);
+
+  // Test both vite and `ng build`
+  await runTests(await ngServe());
+
+  await noSilentNg('build', '--output-mode=server');
+  await runTests(await spawnServer());
+}
+
+const RESPONSE_EXPECTS: Record<
+  string,
+  {
+    matches: RegExp[];
+    notMatches: RegExp[];
+  }
+> = {
+  '/': {
+    matches: [/<link rel="modulepreload" href="(home\.component-[a-zA-Z0-9]{8}\.js)">/],
+    notMatches: [/ssg\.component/, /ssr\.component/, /csr\.component/, /cross-dep-/],
+  },
+  '/ssg': {
+    matches: [
+      /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
+      /<link rel="modulepreload" href="(ssg\.component-[a-zA-Z0-9]{8}\.js)">/,
+    ],
+    notMatches: [
+      /home\.component/,
+      /ssr\.component/,
+      /csr\.component/,
+      /ssg-one\.component/,
+      /ssg-two\.component/,
+      /cross-dep-/,
+    ],
+  },
+  '/ssg/one': {
+    matches: [
+      /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
+      /<link rel="modulepreload" href="(ssg-one\.component-[a-zA-Z0-9]{8}\.js)">/,
+      /<link rel="modulepreload" href="(cross-dep-[a-zA-Z0-9]{8}\.js)">/,
+    ],
+    notMatches: [
+      /home\.component/,
+      /ssr\.component/,
+      /csr\.component/,
+      /ssg-two\.component/,
+      /ssg\.component/,
+    ],
+  },
+  '/ssg/two': {
+    matches: [
+      /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
+      /<link rel="modulepreload" href="(ssg-two\.component-[a-zA-Z0-9]{8}\.js)">/,
+      /<link rel="modulepreload" href="(cross-dep-[a-zA-Z0-9]{8}\.js)">/,
+    ],
+    notMatches: [
+      /home\.component/,
+      /ssr\.component/,
+      /csr\.component/,
+      /ssg-one\.component/,
+      /ssg\.component/,
+    ],
+  },
+  '/ssr': {
+    matches: [/<link rel="modulepreload" href="(ssr\.component-[a-zA-Z0-9]{8}\.js)">/],
+    notMatches: [/home\.component/, /ssg\.component/, /csr\.component/],
+  },
+  '/csr': {
+    matches: [/<link rel="modulepreload" href="(csr\.component-[a-zA-Z0-9]{8}\.js)">/],
+    notMatches: [/home\.component/, /ssg\.component/, /ssr\.component/, /cross-dep-/],
+  },
+};
+
+async function runTests(port: number): Promise<void> {
+  for (const [pathname, { matches, notMatches }] of Object.entries(RESPONSE_EXPECTS)) {
+    const res = await fetch(`http://localhost:${port}${pathname}`);
+    const text = await res.text();
+
+    for (const match of matches) {
+      assert.match(text, match, `Response for '${pathname}': ${match} was not matched in content.`);
+
+      // Ensure that the url is correct and it's a 200.
+      const link = text.match(match)?.[1];
+      const preloadRes = await fetch(`http://localhost:${port}/${link}`);
+      assert.equal(preloadRes.status, 200);
+    }
+
+    for (const match of notMatches) {
+      assert.doesNotMatch(
+        text,
+        match,
+        `Response for '${pathname}': ${match} was matched in content.`,
+      );
+    }
+  }
+}
```

Comment:
I agree that in some cases, overly generic tests can become a source of bugs. I gave refactoring a try, but I found that it made the test harder to maintain. Additionally, since the expectations are no longer combined, it’s more difficult to evaluate and update them when something changes.

In this case, I believe it's better to keep the expectations combined. I encountered this issue myself while working on the feature. For instance, if a new chunk (like cross-dep) is added, It's easier to update all the notMatches, and having all the expectations in one place makes it easy to do that.

Additionally, with the expectations split out, it’s very easy to accidentally use the wrong variable for the response, which could lead to incorrect assertions. IMHO Keeping everything together reduces that risk and ensures clarity.



For context: 
```ts
async function runTests(port: number): Promise<void> {
  // Test for '/'
  const resRoot = await fetch(`http://localhost:${port}/`);
  const textRoot = await resRoot.text();

  const rootMatches = [/<link rel="modulepreload" href="(home\.component-[a-zA-Z0-9]{8}\.js)">/];
  for (const match of rootMatches) {
    assert.match(textRoot, match, `Response for '/': ${match} was not matched in content.`);
    await checkPreloadLink(port, textRoot, match);
  }

  const rootNotMatches = [/ssg\.component/, /ssr\.component/, /csr\.component/, /cross-dep-/];
  for (const match of rootNotMatches) {
    assert.doesNotMatch(textRoot, match, `Response for '/': ${match} was matched in content.`);
  }

  // Test for '/ssg'
  const resSsg = await fetch(`http://localhost:${port}/ssg`);
  const textSsg = await resSsg.text();

  const ssgMatches = [
    /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(ssg\.component-[a-zA-Z0-9]{8}\.js)">/,
  ];
  for (const match of ssgMatches) {
    assert.match(textSsg, match, `Response for '/ssg': ${match} was not matched in content.`);
    await checkPreloadLink(port, textSsg, match);
  }

  const ssgNotMatches = [
    /home\.component/,
    /ssr\.component/,
    /csr\.component/,
    /ssg-one\.component/,
    /ssg-two\.component/,
    /cross-dep-/,
  ];
  for (const match of ssgNotMatches) {
    assert.doesNotMatch(textSsg, match, `Response for '/ssg': ${match} was matched in content.`);
  }

  // Test for '/ssg/one'
  const resSsgOne = await fetch(`http://localhost:${port}/ssg/one`);
  const textSsgOne = await resSsgOne.text();

  const ssgOneMatches = [
    /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(ssg-one\.component-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(cross-dep-[a-zA-Z0-9]{8}\.js)">/,
  ];

  for (const match of ssgOneMatches) {
    assert.match(
      textSsgOne,
      match,
      `Response for '/ssg/one': ${match} was not matched in content.`,
    );
    await checkPreloadLink(port, textSsgOne, match);
  }

  const ssgOneNotMatches = [
    /home\.component/,
    /ssr\.component/,
    /csr\.component/,
    /ssg-two\.component/,
    /ssg\.component/,
  ];

  for (const match of ssgOneNotMatches) {
    assert.doesNotMatch(
      textSsgOne,
      match,
      `Response for '/ssg/one': ${match} was matched in content.`,
    );
  }

  // Test for '/ssg/two'
  const resSsgTwo = await fetch(`http://localhost:${port}/ssg/two`);
  const textSsgTwo = await resSsgTwo.text();

  const ssgTwoMatches = [
    /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(ssg-two\.component-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(cross-dep-[a-zA-Z0-9]{8}\.js)">/,
  ];

  for (const match of ssgTwoMatches) {
    assert.match(
      textSsgTwo,
      match,
      `Response for '/ssg/two': ${match} was not matched in content.`,
    );
    await checkPreloadLink(port, textSsgTwo, match);
  }

  const ssgTwoNotMatches = [
    /home\.component/,
    /ssr\.component/,
    /csr\.component/,
    /ssg-one\.component/,
    /ssg\.component/,
  ];async function runTests(port: number): Promise<void> {
  // Test for '/'
  const resRoot = await fetch(`http://localhost:${port}/`);
  const textRoot = await resRoot.text();
  const rootMatches = [/<link rel="modulepreload" href="(home\.component-[a-zA-Z0-9]{8}\.js)">/];
  for (const match of rootMatches) {
    assert.match(textRoot, match, `Response for '/': ${match} was not matched in content.`);
    await checkPreloadLink(port, textRoot, match);
  }

  assert.doesNotMatch(textRoot, /home\.component/, `Response for '/' did not match`);
  assert.doesNotMatch(textRoot, /ssr\.component/, `Response for '/' did not match`);
  assert.doesNotMatch(textRoot, /csr\.component/, `Response for '/' did not match`);
  assert.doesNotMatch(textRoot, /ssg-two\.component/, `Response for '/' did not match`);
  assert.doesNotMatch(textRoot, /cross-dep-/, `Response for '/' did not match`);

  // Test for '/ssg'
  const resSsg = await fetch(`http://localhost:${port}/ssg`);
  const textSsg = await resSsg.text();
  const ssgMatches = [
    /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(ssg\.component-[a-zA-Z0-9]{8}\.js)">/,
  ];

  for (const match of ssgMatches) {
    assert.match(textSsg, match, `Response for '/ssg': ${match} was not matched in content.`);
    await checkPreloadLink(port, textSsg, match);
  }
  assert.doesNotMatch(textSsg, /home\.component/, `Response for '/ssg' did not match`);
  assert.doesNotMatch(textSsg, /ssr\.component/, `Response for '/ssg' did not match`);
  assert.doesNotMatch(textSsg, /csr\.component/, `Response for '/ssg' did not match`);
  assert.doesNotMatch(textSsg, /ssg-two\.component/, `Response for '/ssg' did not match`);
  assert.doesNotMatch(textSsg, /ssg-one\.component/, `Response for '/ssg' did not match`);
  assert.doesNotMatch(textSsg, /cross-dep-/, `Response for '/ssg' did not match`);

  // Test for '/ssg/one'
  const resSsgOne = await fetch(`http://localhost:${port}/ssg/one`);
  const textSsgOne = await resSsgOne.text();
  const ssgOneMatches = [
    /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(ssg-one\.component-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(cross-dep-[a-zA-Z0-9]{8}\.js)">/,
  ];
  for (const match of ssgOneMatches) {
    assert.match(
      textSsgOne,
      match,
      `Response for '/ssg/one': ${match} was not matched in content.`,
    );
    await checkPreloadLink(port, textSsgOne, match);
  }
  assert.doesNotMatch(textSsgOne, /home\.component/, `Response for '/ssg/one' did not match`);
  assert.doesNotMatch(textSsgOne, /ssr\.component/, `Response for '/ssg/one' did not match`);
  assert.doesNotMatch(textSsgOne, /csr\.component/, `Response for '/ssg/one' did not match`);
  assert.doesNotMatch(textSsgOne, /ssg-two\.component/, `Response for '/ssg/one' did not match`);
  assert.doesNotMatch(textSsgOne, /ssg\.component/, `Response for '/ssg/one' did not match`);

  // Test for '/ssg/two'
  const resSsgTwo = await fetch(`http://localhost:${port}/ssg/two`);
  const textSsgTwo = await resSsgTwo.text();
  const ssgTwoMatches = [
    /<link rel="modulepreload" href="(ssg\.routes-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(ssg-two\.component-[a-zA-Z0-9]{8}\.js)">/,
    /<link rel="modulepreload" href="(cross-dep-[a-zA-Z0-9]{8}\.js)">/,
  ];

  for (const match of ssgTwoMatches) {
    assert.match(
      textSsgTwo,
      match,
      `Response for '/ssg/two': ${match} was not matched in content.`,
    );
    await checkPreloadLink(port, textSsgTwo, match);
  }

  assert.doesNotMatch(textSsgOne, /home\.component/, `Response for '/ssg/two' did not match`);
  assert.doesNotMatch(textSsgOne, /ssr\.component/, `Response for '/ssg/two' did not match`);
  assert.doesNotMatch(textSsgOne, /csr\.component/, `Response for '/ssg/two' did not match`);
  assert.doesNotMatch(textSsgOne, /ssg-one\.component/, `Response for '/ssg/two' did not match`);
  assert.doesNotMatch(textSsgOne, /ssg\.component/, `Response for '/ssg/two' did not match`);

  // Test for '/ssr'
  const resSsr = await fetch(`http://localhost:${port}/ssr`);
  const textSsr = await resSsr.text();
  const ssrMatches = [/<link rel="modulepreload" href="(ssr\.component-[a-zA-Z0-9]{8}\.js)">/];
  for (const match of ssrMatches) {
    assert.match(textSsr, match, `Response for '/ssr': ${match} was not matched in content.`);
    await checkPreloadLink(port, textSsr, match);
  }

  assert.doesNotMatch(textSsgOne, /home\.component/, `Response for '/ssr' did not match`);
  assert.doesNotMatch(textSsgOne, /ssg\.component/, `Response for '/ssr' did not match`);
  assert.doesNotMatch(textSsgOne, /csr\.component/, `Response for '/ssr' did not match`);

  // Test for '/csr'
  const resCsr = await fetch(`http://localhost:${port}/csr`);
  const textCsr = await resCsr.text();
  const csrMatches = [/<link rel="modulepreload" href="(csr\.component-[a-zA-Z0-9]{8}\.js)">/];
  for (const match of csrMatches) {
    assert.match(textCsr, match, `Response for '/csr': ${match} was not matched in content.`);
    await checkPreloadLink(port, textCsr, match);
  }

  assert.doesNotMatch(textSsgOne, /home\.component/, `Response for '/csr' did not match`);
  assert.doesNotMatch(textSsgOne, /ssg\.component/, `Response for '/csr' did not match`);
  assert.doesNotMatch(textSsgOne, /ssr\.component/, `Response for '/csr' did not match`);
  assert.doesNotMatch(textSsg, /cross-dep-/, `Response for '/csr' did not match`);
}

```

Reply:
FWIW, a pattern I like to use for similar repeated code is:

```typescript
it('...', () => {
  // SSR
  {
    const res = await fetch('/ssr');
    expect(res).toBe('...');
  }

  // SSG
  {
    const res = await fetch('/ssg');
    expect(res).toBe('...');
  }

  // CSR
  {
    const res = await fetch('/csr');
    expect(res).toBe('...');
  }
});
```

This scopes each "test case" so you don't accidentally leak state between them.

I'll defer to you on the best way to write this, I can agree it looks quite complicated this way. Just something to consider.